### PR TITLE
Beefy: use VersionedFinalityProof instead of SignedCommitment

### LINKED ISSUE #11838: beefy: SignedCommitment/Justification channels should use versioned struct
There are multiple component sending [BEEFY SignedCommitments](https://github.com/paritytech/substrate/blob/ded44948e2d5a398abcb4e342b0513cb690961bb/primitives/beefy/src/commitment.rs#L142) between them:

- block-import <-> voter channel
- voter <-> RPC-worker channel
- **RPC public API**

All of the above (and possibly others) should instead be using [VersionedFinalityProof](https://github.com/paritytech/substrate/blob/ded44948e2d5a398abcb4e342b0513cb690961bb/primitives/beefy/src/commitment.rs#L290) instead of `SignedCommitment`.

(Changing the definitions of the [notification channels](https://github.com/paritytech/substrate/blob/7409589ab78e7cb508db873dd6384634d667b8f5/client/beefy/src/notification.rs#L37-L42) to use `VersionedFinalityProof` as their payload should surface all the places where this change is required)

### PR DESCRIPTION
**Pull request by hzy1919**
Fixes #11838

And change BEEFY RPC public API . 

polkadot companion: https://github.com/paritytech/polkadot/pull/5852

### CODE REVIEW

File: client/beefy/rpc/src/lib.rs [hunk 1]
```diff
@@ -136,10 +137,10 @@ where
...
 {
 	fn subscribe_justifications(&self, pending: PendingSubscription) {
-		let stream = self
-			.signed_commitment_stream
-			.subscribe()
-			.map(|sc| notification::EncodedSignedCommitment::new::<Block>(sc));
+		let stream = self.signed_commitment_stream.subscribe().map(|vfp| match vfp {
+			VersionedFinalityProof::V1(sc) =>
+				notification::EncodedSignedCommitment::new::<Block>(sc),
+		});
```
Comment (acatangiu) [inline client/beefy/rpc/src/lib.rs:143]: I would actually change the exposed RPCs too. User facing API should support versioning.

File: client/beefy/src/import.rs [hunk 2]
```diff
@@ -116,11 +117,9 @@ where
 		}
 		// Send the justification to the BEEFY voter for processing.
 		match justification {
-			// TODO #11838: Should not unpack, these channels should also use
-			// `VersionedFinalityProof`.
-			VersionedFinalityProof::V1(signed_commitment) => self
+			BeefyVersionedFinalityProof::<Block>::V1(signed_commitment) => self
 				.justification_sender
-				.notify(|| Ok::<_, ()>(signed_commitment))
+				.notify(|| Ok::<_, ()>(BeefyVersionedFinalityProof::<Block>::V1(signed_commitment)))
```
Comment (acatangiu) [inline client/beefy/src/import.rs:122]: no need to unpack here just to pack it back in `V1`, just directly send the `justification`

File: client/beefy/src/worker.rs [hunk 3]
```diff
@@ -417,39 +420,39 @@ where
...
-				self.finalize(signed_commitment);
+				// We created the `finality_proof` and know to be valid.
+				self.finalize(finality_proof);
 			}
 		}
 		Ok(())
 	}
 
-	/// Provide BEEFY finality for block based on `signed_commitment`:
+	/// Provide BEEFY finality for block based on `finality_roof`:
```
Comment (acatangiu) [inline client/beefy/src/worker.rs:444]:
```suggestion
	/// Provide BEEFY finality for block based on `finality_proof`:
```

File: client/beefy/src/worker.rs [hunk 4]
```diff
@@ -417,39 +420,39 @@ where
...
 	/// 1. Prune irrelevant past sessions from the oracle,
 	/// 2. Set BEEFY best block,
-	/// 3. Send best block hash and `signed_commitment` to RPC worker.
+	/// 3. Send best block hash and `finality_roof` to RPC worker.
```
Comment (acatangiu) [inline client/beefy/src/worker.rs:447]:
```suggestion
	/// 3. Send best block hash and `finality_proof` to RPC worker.
```

File: client/beefy/src/worker.rs [hunk 5]
```diff
@@ -417,39 +420,39 @@ where
...
 	///
 	/// Expects `signed commitment` to be valid.
-	fn finalize(&mut self, signed_commitment: BeefySignedCommitment<B>) {
+	fn finalize(&mut self, finality_roof: BeefyVersionedFinalityProof<B>) {
```
Comment (acatangiu) [inline client/beefy/src/worker.rs:450]:
```suggestion
	fn finalize(&mut self, finality_proof: BeefyVersionedFinalityProof<B>) {
```

**Review by acatangiu — COMMENTED** [review]
Changes look good, thank you for contributing!

There's a few nits, and an open question on whether we should also change public facing RPC API (breaking change). I think we have to do it eventually if we really want to support versioned proofs, and better to do it now before deploying to production networks.

File: client/beefy/rpc/src/lib.rs (hunk 1 above)
Comment (hzy1919) [reply client/beefy/rpc/src/lib.rs:143]: I'll change the exposed RPCs now.

File: client/beefy/src/import.rs (hunk 2 above)
Comment (hzy1919) [reply client/beefy/src/import.rs:122]: Right, I made confused here. :smile:

**Review by svyatonik — APPROVED** [review]
LGTM

**Review by acatangiu — APPROVED** [review]
Looks good, thanks!

Please add to the PR description the fact that BEEFY RPC API changed and should be included in release notes.

**Comment by hzy1919** [comment]
> Looks good, thanks!
> 
> Please add to the PR description the fact that BEEFY RPC API changed and should be included in release notes.

Ok .  Done.

**Comment by acatangiu** [comment]
@hzy1919 thanks for contributing! you can add a polkadot address for a tip

**Comment by acatangiu** [comment]
bot merge

**Comment by hzy1919** [comment]
> @hzy1919 thanks for contributing! you can add a polkadot address for a tip

@acatangiu You're welcome .  I have added my polkadot address in the PR description.

**Pull request merged**